File: mathewDavidov/assignments/reverseLinked/lc92/lc92.java
/**
 * Problem Statement:
 * Given the head of a LinkedList and two positions ‘p’ and ‘q’, 
 * reverse the LinkedList from position ‘p’ to ‘q’.
 * 
 * Example 1:
 * 1 -> 2 -> 3 -> 4 -> 5
 * p = 2, q = 4
 * 
 * 1 -> 4 -> 3 -> 2 -> 5
 * 
 * Edge cases:
 * null head / one element -> return
 * p comes after q -> return head
 * 
 * initialize a dummy node to before the beginning of the list (to have a reference at the end to return)
 * loop from 0 to p - 1
 *   move the pointer forward (this is where we reverse the list from)
 * reverse the list:
 *   initialize a reference to the reverse node
 *   initialize a reference to the current node (previous' next)
 *   loop from 0 to q - p + 1 times:
 *     // (follow the same logic as 206)
 *     initialize a reference to the next node to the current's next node
 *     assign the current's next node to the reverse node
 *     point the reverse node to the current node
 *     point the current node to the next node
 *   point the previous' next next to current
 *   point the previous next to reverse
 *   return the head
 * 
 * Runtime:
 * Time: O(n)
 * Space: O(1)
 */

public class lc92 {
    public static ListNode reverseBetween(ListNode head, int p, int q) {
        if (head == null || head.next == null || p > q) {
            return head;
        }

        ListNode beforeList = new ListNode(0);
        beforeList.next = head;
        // e.g. before -> 1 -> 2 -> 3

        ListNode previous = beforeList;

        for (int i = 0; i < p - 1; i++) {
            previous = previous.next;
        }

        ListNode current = previous.next;
        ListNode reverse = null;
        
        for (int i = 0; i < q - p + 1; i++) {
            ListNode next = current.next;
            current.next = reverse;
            reverse = current;
            current = next;
        }
        
        previous.next.next = current;
        previous.next = reverse;

        return beforeList.next;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);

        ListNode result = reverseBetween(head, 2, 4);
        System.out.println("Nodes of the reversed LinkedList are: ");
        while (result != null) {
            System.out.println(result.value);
            result = result.next;
        }
    }
}

class ListNode {
    public int value = 0;
    ListNode next;

    ListNode(int value) {
        this.value = value;
        next = null;
    }
}

File: AdrianBarros/assignments/reverseLinked/lc92/lc92.py
# Problem Statement #

# Given the head of a LinkedList and two positions

from __future__ import print_function


class Node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next

    def print_list(self):
        temp = self
        while temp is not None:
            print(temp.value, end=" ")
            temp = temp.next
        print()


'''
input: singly connected linklist, 2 integers
output: head othe same linklist

step 1: check if there is linklist, if num1 < num2
step 2: loop throught the link list prev, curr and next until we hit the first node
Added step: keep track of both the last node before the reverse and the last node of the sub LL
added step: separeate the sub LL
step 3: while loop reversing the link list until we hit the second node
step 4: reconnect the sub LL
final step: return the head of the original LL

edge cases:
go to step 1
if given node == head
'''


def reverse_sub_list(head, p, q):
    # TODO: Write your code here
    if p == q:
        return head
    # declaring ptrs
    prev = None
    curr = head
    next = head
    # step 2: loop throught the link list prev, curr and next until we hit the first node
    while curr.value != p:
        prev = curr
        curr = curr.next
    # we have to keep track of the last node
    last_node = prev
    new_start = curr
    # step 3: while loop reversing the link list until we hit the second node

    i = 0
  # reverse nodes between 'p' and 'q'
    while curr is not None and i < q - p + 1:
        next = curr.next
        curr.next = prev
        prev = curr
        curr = next
        i += 1

    if last_node is not None:
        # 'previous' is now the first node of the sub-list
        last_node.next = prev
  # this means p == 1 i.e., we are changing the first node (head) of the LinkedList
    else:
        head = prev
        # connect with the last part
    new_start.next = curr
    return head


def main():

    head = Node(1)  # curr
    head.next = Node(2)
    head.next.next = Node(3)
    head.next.next.next = Node(4)
    head.next.next.next.next = Node(5)

    print("Nodes of original LinkedList are: ", end='')
    head.print_list()
    result = reverse_sub_list(head, 2, 4)
    print("Nodes of reversed LinkedList are: ", end='')
    result.print_list()


main()


# Solution
# -----
# def reverse_sub_list(head, p, q):
#   if p == q:
#     return head

#   # after skipping 'p-1' nodes, current will point to 'p'th node
#   current, previous = head, None
#   i = 0
#   while current is not None and i < p - 1:
#     previous = current
#     current = current.next
#     i += 1

#   # we are interested in three parts of the LinkedList, the part before index 'p',
#   # the part between 'p' and 'q', and the part after index 'q'
#   last_node_of_first_part = previous
#   # after reversing the LinkedList 'current' will become the last node of the sub-list
#   last_node_of_sub_list = current
#   next = None  # will be used to temporarily store the next node

#   i = 0
#   # reverse nodes between 'p' and 'q'
#   while current is not None and i < q - p + 1:
#     next = current.next
#     current.next = previous
#     previous = current
#     current = next
#     i += 1

#   # connect with the first part
#   if last_node_of_first_part is not None:
#     # 'previous' is now the first node of the sub-list
#     last_node_of_first_part.next = previous
#   # this means p == 1 i.e., we are changing the first node (head) of the LinkedList
#   else:
#     head = previous

#   # connect with the last part
#   last_node_of_sub_list.next = current
#   return head

# -----
